### ISSUE TITLE
2x Warehouse Nodes Stopped Replaying and Uploading Blocks But Remained Online

### ISSUE THREAD

**Issue by buffalu** [solana-labs/solana#29901]
#### Problem
Our two warehouse nodes running v1.14.13-jito stopped processing slots last night. They were still online, but weren't able to progress. One stopped around 3AM CT and another stopped around 7:45AM CT.

They also stopped uploading to bigtable

Identities:
- 5oXsXDXM8kmcK3KjSpurxB8DsJUqWCrSiD2kcLWEniSy
- ATb9AnWw12SsUGtG4LadUhHL276Tf51u1XTmt8kFY3iG

#### Proposed Solution
Determine why they stopped freezing slots and fix

**Comment by buffalu** [comment]
![Screen Shot 2023-01-25 at 9 48 40 AM](https://user-images.githubusercontent.com/85544055/214609454-d6e72ec0-5aef-4662-976e-c4c2347ebcb4.png)

**Comment by buffalu** [comment]
bigtable slot upload chart

![Screen Shot 2023-01-25 at 9 55 51 AM](https://user-images.githubusercontent.com/85544055/214611410-e5b8dc55-7778-4ebd-a91d-4396d4761c90.png)

**Comment by steviez** [comment]
I believe bigtable will only upload rooted slots. It is possible that banks were getting frozen but not rooted (ie a consensus divergence); can you check your logs or metrics to see if your nodes were freezing banks?

**Comment by buffalu** [comment]
it happened again

logs: https://drive.google.com/file/d/1WVpB8pjVoUHg7WWlWdtd4uggE8SmE_fv/view?usp=share_link

bigtable upload (times are central time, jan 25 2023)
22:39:24 was last block uploaded

<img width="1621" alt="Screen Shot 2023-01-25 at 11 01 29 PM" src="https://user-images.githubusercontent.com/85544055/214762017-342a2685-3dd2-4d7d-8b99-ec87f7c21542.png">

**Comment by buffalu** [comment]
last frozen bank:
```
Jan 26 04:39:35 ny-mainnet-warehouse-1 solana-validator[2643753]: [2023-01-26T04:39:35.044682103Z INFO  solana_runtime::bank] bank frozen: 174429976 hash: 2Fn27poQeRZUqbuyQgFPvzqjkReuYH2SDnAkupofpFuC accounts_delta: FUg7UnKSqy4K4W4TBCkHx8y6pPfamFWNU2LfTfdu3NWy signature_count: 2524 last_blockhash: 2avz5dUychyLytLUCpcpDHpYP16fMC6JEeHHYadRuwmw capitalization: 539312136320050691
```

**Comment by buffalu** [comment]
looks like a total deadlock? this is replay-loop-timing-stats which seems like it should be emitted every iteration with 1s rate limit

<img width="1607" alt="Screen Shot 2023-01-25 at 11 14 47 PM" src="https://user-images.githubusercontent.com/85544055/214763340-f2dbc527-aa7d-43f6-a23e-3bc67fadc6ec.png">

**Comment by steviez** [comment]
Taking a quick look at your log, I would agree that deadlock seems like a possibility. I think getting some stracktraces would be helpful to figure out where things are stuck. General process:
1. Rebuild binary with debug symbols
```
RUSTFLAGS='-g -C force-frame-pointers=yes' cargo build --release --bin solana-validator
```
2. Restart process and wait
3. Wait for hang again
4. Attach `gdb` and dump backtrace for all threads (the `(gdb)` lines are run within gdb)
```
$ gdb -p <PID>
(gdb) set logging file ~/gdb.log
(gdb) set logging on
(gdb) thread apply all bt full
(gdb) quit
$ cat ~/gdb.log
```
With the log, you should be able to inspect the backtraces and see which threads are stuck.

**Comment by segfaultdoc** [comment]
> Taking a quick look at your log, I would agree that deadlock seems like a possibility. I think getting some stracktraces would be helpful to figure out where things are stuck. General process:
> 
> 1. Rebuild binary with debug symbols
> 
> ```
> RUSTFLAGS='-g -C force-frame-pointers=yes' cargo build --release --bin solana-validator
> ```
> 
> 2. Restart process and wait
> 3. Wait for hang again
> 4. Attach `gdb` and dump backtrace for all threads (the `(gdb)` lines are run within gdb)
> 
> ```
> $ gdb -p <PID>
> (gdb) set logging file ~/gdb.log
> (gdb) set logging on
> (gdb) thread apply all bt full
> (gdb) quit
> $ cat ~/gdb.log
> ```
> 
> With the log, you should be able to inspect the backtraces and see which threads are stuck.

Trying to get gdb installed on a package manager-less system is a much more daunting task than I would have imagined. In the meantime I downgraded one of the nodes to 1.13.6 and seeing the same symptoms of no bigtable uploads

**Comment by kunal365roy** [comment]
We've seen issues similar to this and have founding removing the accounts, ledger and snapshots directory and restarting while fetching the snapshot will reestablish bt connection + upload reliably. I was hoping this patch in 1.13.6 would stop the bt upload sync https://github.com/solana-labs/solana/pull/28728

**Comment by steviez** [comment]
> We've seen issues similar to this

Did your node completely lock up (ie stop replaying slot) or did your node just stop uploading to bigtable while continuing to remain with the tip of the cluster? The issue you linked sounds like a session issue (ie the latter) whereas this issue is exploring a fundamental issue where the validator client grinds to a halt across the board.

> removing the accounts, ledger and snapshots directory and restarting while fetching the snapshot will reestablish bt connection + upload reliably

This is a bandaid, and should really be a last resort. Did you ever try stopping/restarting your nodes along the way without wiping everything? If you're still encountering this, do you currently have any open issues and/or PR's ?

Additionally, if you're uploading to bigtable, suppose:
- Your nodes last uploads at slot `X`
- You then wipe your local state and restart
- Upon restart, your node fetches a snapshot at slot `Y`

You now have a hole in your history from `(X, Y]` ...

**Comment by segfaultdoc** [comment]
Here's a gdb stacktrace of our locked up node
[
[gdb.log](https://github.com/solana-labs/solana/files/10580625/gdb.log)
](url)

**Comment by kunal365roy** [comment]
> Did your node completely lock up (ie stop replaying slot) or did your node just stop uploading to bigtable while continuing to remain with the tip of the cluster? The issue you linked sounds like a session issue (ie the latter) whereas this issue is exploring a fundamental issue where the validator client grinds to a halt across the board.

Ah I read too fast apologize, we also have this as a separate issue. I'm referring to the one where the validator continues to sink but loses the bt session, fixed by the latest release.

My validators have been very inconsistently syncing as of late 


> Additionally, if you're uploading to bigtable, suppose:
> Your nodes last uploads at slot X
> You then wipe your local state and restart
> Upon restart, your node fetches a snapshot at slot Y
> You now have a hole in your history from (X, Y] ...

Yeah we have an alerter and backfill async job that fills that with the solana ledger tool

**Comment by segfaultdoc** [comment]
Adding to this thread that 1.13.5 ran into the issue, bumping down to 1.13.4

**Comment by steviez** [comment]
> Here's a gdb stacktrace of our locked up node [ [gdb.log](https://github.com/solana-labs/solana/files/10580625/gdb.log) ](url)

Hi @segfaultdoc - the log you posted appears to be truncated a bit, I don't see all threads; any chance you saved the output and have the full log available?

**Comment by steviez** [comment]
Thanks for the updated gdb output. It looks like some of the threads are still missing, but there is output for more threads from this second output which help shed some light. It looks like some threads are stuck on this lock:
https://github.com/solana-labs/solana/blob/d53c49c230d9b6cd464113e8b7cc6798b4301431/ledger/src/blockstore.rs#L181

I see `solana-ledger-p` (`solana-ledger-purge`) thread waiting on this lock:
https://github.com/solana-labs/solana/blob/46fde27b3dd95a4d06020150ab3e1c0e93746db7/core/src/ledger_cleanup_service.rs#L240

```
Thread 967 (Thread 0x7ea9d1d1a640 (LWP 1010644) "solana-ledger-p"):
#0  syscall () at ../sysdeps/unix/sysv/linux/x86_64/syscall.S:38
No locals.
#1  0x000055a9efa6a825 in std::sys::unix::futex::futex_wait () at library/std/src/sys/unix/futex.rs:62
No locals.
#2  0x000055a9edb68602 in std::sys::unix::locks::futex_rwlock::RwLock::write_contended () at library/std/src/sys/unix/locks/futex_rwlock.rs:220
No locals.
#3  0x000055a9ee0e15e3 in std::sys::unix::locks::futex_rwlock::RwLock::write (self=0x7f02562e5138) at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/sys/unix/locks/futex_rwlock.rs:156
No locals.
#4  std::sys_common::rwlock::MovableRwLock::write (self=0x7f02562e5138) at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/sys_common/rwlock.rs:102
No locals.
#5  std::sync::rwlock::RwLock<u64>::write<u64> (self=0x7f02562e5138) at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/sync/rwlock.rs:298
No locals.
#6  solana_core::ledger_cleanup_service::{impl#0}::cleanup_ledger::{closure#0} () at core/src/ledger_cleanup_service.rs:240
```

Several `bigtable-upload` threads are also waiting on the same lock (through a series of function calls):
https://github.com/solana-labs/solana/blob/46fde27b3dd95a4d06020150ab3e1c0e93746db7/ledger/src/blockstore.rs#L2576

```
Thread 783 (Thread 0x7eac2d4d0640 (LWP 1010407) "bigtable-upload"):
#0  syscall () at ../sysdeps/unix/sysv/linux/x86_64/syscall.S:38
No locals.
#1  0x000055a9efa6a825 in std::sys::unix::futex::futex_wait () at library/std/src/sys/unix/futex.rs:62
No locals.
#2  0x000055a9edb684cc in std::sys::unix::locks::futex_rwlock::RwLock::read_contended () at library/std/src/sys/unix/locks/futex_rwlock.rs:139
No locals.
#3  0x000055a9ee85750f in std::sys::unix::locks::futex_rwlock::RwLock::read (self=0x7f02562e5138) at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/sys/unix/locks/futex_rwlock.rs:88
        state = <optimized out>
        state = <optimized out>
#4  std::sys_common::rwlock::MovableRwLock::read (self=0x7f02562e5138) at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/sys_common/rwlock.rs:86
No locals.
#5  std::sync::rwlock::RwLock<u64>::read<u64> (self=0x7f02562e5138) at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/sync/rwlock.rs:210
No locals.
#6  solana_ledger::blockstore::Blockstore::check_lowest_cleanup_slot (self=0x7f02562e5010, slot=175822617) at ledger/src/blockstore.rs:2576
No locals.
#7  0x000055a9ee85d3ca in solana_ledger::blockstore::Blockstore::get_completed_ranges (self=0x89, slot=3221225478, start_index=139648018373181) at ledger/src/blockstore.rs:3193
No locals.
#8  0x000055a9ee85ce73 in solana_ledger::blockstore::Blockstore::get_slot_entries_with_shred_info (self=0x89, slot=175822617, start_index=0, allow_dead_slots=false) at ledger/src/blockstore.rs:3150
No locals.
#9  0x000055a9ee855db0 in solana_ledger::blockstore::Blockstore::get_slot_entries (self=0x7f02562e5010, shred_start_index=0, slot=<optimized out>) at ledger/src/blockstore.rs:3138
No locals.
```

We do `.unwrap()` when we attempt to grab the locks; these callsites didn't panic so the lock seemingly is NOT poisoned. Also, I didn't see any mention of panics from y'all, but might be a worthwhile sanity check real quick.

The full output could be helpful in narrowing down the search space, altho there should only be a handful of threads that grab a writer on that lock so we might be able to proceed without it as well.

**Comment by steviez** [comment]
Got back to this, looks like we have a call-path that will attempt to obtain the a read-lock a second time while it is already held. The callpath is:
```
+ Call Blockstore::get_rooted_block()
+!!! Read lock obtained (line 1926) via Blockstore:: check_lowest_cleanup_slot
+--+ Call Blockstore:: get_complete_block() (line 1929)
   +--+ Call Blockstore::get_slot_entries() (line 1948)
      +--+ Call Blockstore::get_slot_entries_with_shred_info() (line 2778)
         +--+ Call Blockstore::get_completed_ranges() (line 2790)
            +!!! Read lock obtained (line 2856) via Blockstore:: check_lowest_cleanup_slot()
```

https://github.com/solana-labs/solana/blob/bfc29a6b2fd4fca57950f74cce412614f02c22ef/ledger/src/blockstore.rs#L1920-L1932
https://github.com/solana-labs/solana/blob/bfc29a6b2fd4fca57950f74cce412614f02c22ef/ledger/src/blockstore.rs#L1934-L2017
https://github.com/solana-labs/solana/blob/bfc29a6b2fd4fca57950f74cce412614f02c22ef/ledger/src/blockstore.rs#L2777-L2780
https://github.com/solana-labs/solana/blob/bfc29a6b2fd4fca57950f74cce412614f02c22ef/ledger/src/blockstore.rs#L2784-L2832
https://github.com/solana-labs/solana/blob/bfc29a6b2fd4fca57950f74cce412614f02c22ef/ledger/src/blockstore.rs#L2860-L2882

I'll look at re-arranging stuff to avoid this and throw a PR up

**Comment by steviez** [comment]
@buffalu @segfaultdoc - The proposed fix is a little scary given that we're playing with locks, so we just want to be extra sure that there isn't anything else at play.

Could you please confirm whether this warehouse node is doing anything else (ie serving RPC requests)? Additionally, can you comment on CPU utilization before/during the observed hang?
